File: Banking System PY/AcCreation.py
import json
import os
import hashlib
from datetime import datetime
class AcCreation:
    acc_storage = "acc.json"
    
    def __init__(self):
        self.setName()
        self.setSurname()
        self.setUsername()
        self.setPin()
        self.setPhoneNum()
        self.setInitDeposit()
        self.__transactions = []
        
        
    def setName(self):
        hold = True
        while(hold):
            name = input("please enter a valid name: ")
            if name.isalpha():
                hold = False
        self.__name = name
        
    def setSurname(self):
        hold = True
        while(hold):
            surname = input("please enter a valid surname: ")
            if surname.isalpha():
                hold = False
        self.__surname = surname
        
    def setUsername(self):
        hold = True 
        while hold:
            username = input("enter your desired username: ")
            if self.isUsernameUnique(username):
                choice = input("are you sure you want to proceed with this username? (enter 'yes' or 'no') ")
                if choice.lower() =="yes":
                    self.__username = username
                    hold = False
            else:
                print("username taken, choose another one")

    def setPin(self):
        hold = True
        while hold:
            pin = input("enter your desired pin, atleast 4 digits: ")
            if pin.isdigit() and len(pin)>=4:
                self.__pin = pin
                hold = False
                         
    def setPhoneNum(self):
        hold = True
        while(hold):
            phoneNum = input("Please enter a valid phone number: ")
            if phoneNum.startswith('0') and phoneNum.isdigit() and '-' not in phoneNum:
                hold = False
            elif phoneNum.isdigit() and '-' not in phoneNum:
                hold = False
                
        self.__phoneNum = phoneNum
        
    def setInitDeposit(self):
        self.__balance = 0
        hold = True
        while(hold):
            balance = input("Please enter your initial deposit if there is any: ")
            if balance.isdigit() and '-' not in balance:
                hold = False
        self.__balance = int(balance)
    
    def addTransaction(self,transaction_type,amount):
        transaction={
            "type": transaction_type,
            "amount": amount,
            "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        }
        self.__transactions.append(transaction)
        
        
    def getNameAndSurname(self):
        return self.__name +" "+self.__surname
    
    def getPhoneNum(self):
        return self.__phoneNum
    
    def getBalance(self):
        return self.__balance
    
    def getUsername(self):
        return self.__username
    
    def getPin(self):
        return self.__pin
    '''
    def AccInformations(self):
        hold = True
        while hold:
            if self.verifyPin():
                print(f"Name and Surname: {self.getNameAndSurname()}")
                print(f"Phone number: {self.getPhoneNum()}")
                print(f"Balance: {self.getBalance()}")
                print(f"Username: {self.getUsername()}")
                print(f"pin: {self.getPin()}")
                hold = False
            else:
                print("Failed to verify PIN")
    '''
    def isUsernameUnique(self, username):
        if os.path.exists(self.acc_storage):
            with open(self.acc_storage, "r") as fl:
                try:
                    accounts = json.load(fl)
                    return all(acc["Username"] != username for acc in accounts)
                except json.JSONDecodeError:
                    return True
        return True
    '''
    def verifyPin(self):
        for i in range(3): #3 attempts
            check = input("Enter your PIN: ")
            if check == self.getPin():
                return True
            print("incorrect PIN. Try again.")
        print("Too many incorrect attempts. Exiting.")
        return False
    '''
        
    def getDict(self):
        #convert the object to a python to a dictionary suitable for JSON storage
        return {
            "Name": self.__name,
            "Surname": self.__surname,
            "Phone number": self.__phoneNum,
            "Balance": self.__balance,
            "Username": self.__username,
            "Pin": self.__pin,
            "Transactions": self.__transactions    
            }
        
    def saveToJson(self):
        try:
            if os.path.exists(self.acc_storage):
                with open(self.acc_storage, "r") as fl:
                    try:
                        accounts = json.load(fl)
                    except json.JSONDecodeError:
                        accounts = []
            else:
                accounts = []
                
            accounts.append(self.getDict())
            
            with open(self.acc_storage,"w") as fl:
                json.dump(accounts,fl,indent = 3)
        except Exception as e:
            print(f"error occured: {e}")
        


        
        
    
                
                

    
    

File: Banking-System-JS/AcCreation.js
const fs = require("fs");
const readlineSync = require("readline-sync");

class AcCreation {
  static accStorage = "acc.json";

  constructor() {
    try {
      this.name = this.setName();
      this.surname = this.setSurname();
      this.username = this.setUsername();
      this.pin = this.setPin();
      this.phoneNumber = this.setPhoneNumber();
      this.initDeposit = this.setInitDeposit();
      this.Balance = parseInt(this.initDeposit);
      this.transactions = [];
    } catch (error) {
      console.error("Error during account creation:", error.message);
      throw error;
    }
  }

  setName() {
    while (true) {
      try {
        const name = readlineSync.question("Please enter your name: ");
        if (/^[a-zA-Z]+$/.test(name)) return name;
        throw new Error("Invalid name format");
      } catch (error) {
        console.error(error.message);
      }
    }
  }

  setSurname() {
    while (true) {
      try {
        const surname = readlineSync.question("Please enter your surname: ");
        if (/^[a-zA-Z]+$/.test(surname)) return surname;
        throw new Error("Invalid surname format");
      } catch (error) {
        console.error(error.message);
      }
    }
  }

  setUsername() {
    while (true) {
      try {
        const username = readlineSync.question("Please enter your username: ");
        if (this.isUsernameUnique(username) && username.length < 12) return username;
        throw new Error("Username taken or too long");
      } catch (error) {
        console.error(error.message);
      }
    }
  }

  setPin() {
    while (true) {
      try {
        const pin = readlineSync.question("Please enter your pin (at least 4 digits): ");
        if (pin.length >= 4 && /^\d+$/.test(pin)) return pin;
        throw new Error("Invalid PIN format");
      } catch (error) {
        console.error(error.message);
      }
    }
  }

  setPhoneNumber() {
    while (true) {
      try {
        const phoneNumber = readlineSync.question("Please enter your phone number: ");
        if (/^\d{7,}$/.test(phoneNumber)) return phoneNumber;
        throw new Error("Invalid phone number format");
      } catch (error) {
        console.error(error.message);
      }
    }
  }

  setInitDeposit() {
    while (true) {
      try {
        const initDeposit = readlineSync.question("Please enter your initial deposit: ");
        if (/^\d+$/.test(initDeposit)) return parseInt(initDeposit);
        throw new Error("Invalid initial deposit format");
      } catch (error) {
        console.error(error.message);
      }
    }
  }

  toJson() {
    try {
      return {
        Name: this.name,
        Surname: this.surname,
        Username: this.username,
        Pin: this.pin,
        PhoneNumber: this.phoneNumber,
        Balance: this.Balance,
        Transactions: this.transactions,
      };
    } catch (error) {
      console.error("Error creating JSON representation:", error.message);
      throw error;
    }
  }

  saveToJson() {
    try {
      let accounts = [];
      
      
      if (fs.existsSync(AcCreation.accStorage)) {
        const fileContent = fs.readFileSync(AcCreation.accStorage, "utf8");
        if (fileContent.trim() !== "") {
          try {
            accounts = JSON.parse(fileContent);
          } catch (parseError) {
            console.error("Error parsing JSON:", parseError.message);
            accounts = [];  
          }
        }
      }

      accounts.push(this.toJson());
      fs.writeFileSync(AcCreation.accStorage, JSON.stringify(accounts, null, 4));
    } catch (error) {
      console.error("Error saving account to JSON file:", error.message);
      throw error;
    }
  }

  isUsernameUnique(username) {
    try {
      if (fs.existsSync(AcCreation.accStorage)) {
        const fileContent = fs.readFileSync(AcCreation.accStorage, "utf8");
        if (fileContent.trim() === "") return true;
        const accounts = JSON.parse(fileContent);
        return !accounts.some((acc) => acc.Username === username);
      }
      return true;
    } catch (error) {
      console.error("Error checking username uniqueness:", error.message);
      return true;
    }
  }

  addTransaction(transactionType, amount) {
    try {
      const transaction = {
        type: transactionType,
        amount,
        time: new Date().toISOString(),
      };
      this.transactions.push(transaction);
    } catch (error) {
      console.error("Error adding transaction:", error.message);
      throw error;
    }
  }
}

module.exports = AcCreation;